Change error domains on @objc enums to ensure they're unique

beccadax opened the pull request:
Consider this code:

```swift
class GithubAccount: NSObject {
  @objc(GithubSyncError) enum SyncError: Int, Error { ... }
}
class BitbucketAccount: NSObject {
  @objc(BitbucketSyncError) enum SyncError: Int, Error { ... }
}
```

Because they are @objc enums, Swift automatically derives error domain strings for these two error types of the form "ModuleName.SyncError", ignoring the fact that they are nested in different types and have different Objective-C names. That means an error created in Objective-C with `[[NSError alloc] initWithDomain:GithubSyncErrorDomain code:… userInfo:…]` can be interpreted as a `BitbucketAccount.SyncError` in Swift and vice versa.

(Thankfully, this can't happen if you cast an error created in Swift to `NSError` and then back to the other type.)

This pull request generates domain strings more robustly:

* There is now one function which both PrintAsObjC and @objc enum error derivation use to generate their error domain strings. It doesn't support all types, but it does support everything PrintAsObjC should include.
* This function now includes the parents of nested types.
* If a type is not within the subset supported by the function, the compiler will derive an implementation which calls `String(reflecting:)` on the error type at runtime.

The result is that all Swift error types (except those which have customized their error domains) should have error domain strings equal to calling `String(reflecting:)` on the type, but if there are any bugs in the algorithm, the error domain constant in the generated Objective-C header will match the one seen at runtime. Error domains for most types will not change, but error domains for nested types will; this should only matter in code which has hardcoded those domain strings instead of using the constants.

Resolves [SR-9389](https://bugs.swift.org/browse/SR-9389) and rdar://problem/46420910.

beccadax committed:
[NFC] Extract helper for making NSError domains

Avoids having to coordinate changes in two different parts of the compiler.

brentdax commented:
@swift-ci please test source compatibility

brentdax commented on test/stdlib/ErrorBridged.swift:
```diff
@@ -722,4 +722,38 @@ ErrorBridgingTests.test("Error archetype identity") {
+
+  // Identically-named children of different parents should not be castable to
+  // one another.
+  expectTrue(ParentA.Error.failed as NSError is ParentA.Error)
```
I just realized that these last four tests are written in a form that wouldn't fail in master. I'll fix it tomorrow.

jrose-apple commented on lib/AST/SwiftNameTranslation.cpp:
```diff
@@ -51,6 +52,14 @@ getNameForObjC(const ValueDecl *VD, CustomNamesOnly_t customNamesOnly) {
+getErrorDomainStringForObjC(const EnumDecl *ED) {
+  std::string buffer = ED->getParentModule()->getNameStr();
+  buffer += ".";
+  buffer += ED->getNameStr();
```
This didn't actually change any behavior. Did you mean to use the `@objc` name? Note that not every `@objc` enum *has* a custom `@objc` name, though.

The fact that we *do* print nested enums makes this trickier than it otherwise would be.

brentdax replied:
The first commit refactored the code; [the second changed the behavior](https://github.com/apple/swift/pull/21452/commits/152158ea974df33eff8dd13c904d7c35d30668fb#diff-b62ed7b05c464546f0a07d1e285c96d9R57).

jrose-apple replied:
Oh, bah, I must have just been looking at the first one. Nevertheless, the new behavior isn't good enough either:

```swift
class Foo {
  @objc enum Error: Swift.Error { case x }
}

class Bar {
  @objc enum Error: Swift.Error { case x }
}
```

beccadax committed:
Use more robust Swift type names for error domains

getErrorDomainStringForObjC() now includes the parent types in the error domain string. Its implementation does not support generic types or private discriminators, but those types can’t be PrintAsObjC’d anyway, so we should never see them.

brentdax commented:
After sleeping on it, I decided that people have probably hardcoded error domains in various places and we don't want to break them unnecessarily. So today's commits go back to using the Swift name of the type and simply generate it more robustly; that way only nested error types will change domains. `getErrorDomainStringForObjC()` doesn't handle everything that can come up in a Swift type, but it handles everything that should come up in a PrintAsObjC-able type; if the compiler detects that a given type has a construct it can't handle, it derives a call to `String(reflecting: self)` instead.

The version history here is now dirty enough that I might rewrite history in the morning, but I think this is the right solution.

brentdax commented:
@swift-ci please test

jrose-apple commented on lib/Sema/DerivedConformanceError.cpp:
```diff
@@ -27,12 +27,38 @@
 
-static void deriveBodyBridgedNSError_enum_nsErrorDomain(
-              AbstractFunctionDecl *domainDecl, void *) {
+static void deriveBodyUsingRuntime(AbstractFunctionDecl *domainDecl, void *) {
```
Nitpick: derive body of what? (The old name is a hierarchical "derivedBody\<Protocol>\_\<kind>\_\<requirement>".)

brentdax replied:
I know. With all of that *and* the `UsingRuntime`/`UsingObjcTranslation` suffix, the function name was nearing the line length limit on its own. I figured I could drop that information because the file only derives one requirement's body anyway, but I can just use the giant name if you'd prefer.

brentdax replied:
It's more "objcPrintableEnum", but that could work.

jrose-apple commented on lib/Sema/DerivedConformanceError.cpp:
```diff
@@ -27,12 +27,38 @@
+  domainDecl->setBody(body);
+}
+
+static void deriveBodyUsingObjcTranslation(AbstractFunctionDecl *domainDecl,
```
Nitpick: "ObjC"

jrose-apple reviewed (commented):
Looking good!

beccadax committed:
Only generate domains for PrintAsObjC-able types

Otherwise we generate a call to String(reflecting:), which correctly handles many things we may not be able to (like private types), and which matches the default implementation of Error._domain.

brentdax commented:
Rewrote history to remove intermediate states.

brentdax commented:
@swift-ci please test

brentdax commented:
Updated the commit description to reflect the current solution. @kballard, as the person who reported both SR-9389 and SR-700, does this approach look good to you?

(There's no rush; this won't merge until the new year.)

kballard commented on lib/Sema/DerivedConformanceError.cpp:
```diff
@@ -22,11 +22,40 @@
+  // enum SomeEnum {
+  //   @derived
+  //   static var _nsErrorDomain: String {
+  //     return _typeName(self, qualified: true)
```
Should this not be `return String(reflecting: self)`?

brentdax replied:
Ack, you're right. Artifact of this PR's convoluted history.

kballard commented on lib/Sema/DerivedConformanceError.cpp:
```diff
@@ -22,11 +22,40 @@
+                    AbstractFunctionDecl *domainDecl, void *) {
   // enum SomeEnum {
   //   @derived
   //   static var _nsErrorDomain: String {
```
This comment is now out of date. It claims that this returns `"ModuleName.SomeEnum"` but that's no longer true.

brentdax replied:
That's still true for the example given, since it's not shown to be nested. And in any case, the comment is primarily meant to illustrate the shape of the code we're generating (i.e. that we're generating an accessor which returns a string literal) rather than the contents of the string.

lilyball reviewed (commented):
This approach looks fine to me. The fact that `@objc` enums nested in private types don't ever make it to the Obj-C header¹ allays my concerns about having the domain end up including the private discriminators.

¹I actually didn't realize this was true. Testing in Swift 4.2, it looks like nested types only show up in the Obj-C header if their parent type is also in the Obj-C header, which makes sense but wasn't obvious. I wonder if we should have a lint that flags `@objc` attributes on types that won't actually work due to this rule.

brentdax commented:
@swift-ci please smoke test and merge

beccadax committed:
Fix out-of-date comment

brentdax commented:
@swift-ci please smoke test and merge

brentdax commented:
@swift-ci please smoke test
